libminifi/include/core/controller/ForwardingControllerServiceProvider.h

```diff
@@ -0,0 +1,136 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <string>
+
+#include "ControllerServiceProvider.h"
+#include "ControllerServiceNode.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace core {
+namespace controller {
+
+class ForwardingControllerServiceProvider : public ControllerServiceProvider {
```

Comment:
Why do we need pimpl here?

At first look I'm not sure I perfectly understand the reason of introducing this class.

Reply:
its purpose is to make the `FlowController` class skinnier as I was finding myself scrolling a lot, and I thought it is a general-enough behavior to factor out

Reply:
I agree with the general purpose (make flowcontroller tinier), but not sure if pimpl is the best solution to achieve this. Usually introduces a lot of boilerplate, so I'm against it unless we can prove that we really need it. 

If we don't have any better idea, let's do it, but I hope that we will. @szaszm @fgerlits what's your opinion?

Reply:
note that it was already present in `FlowController`, it has just been moved out, I am all for eliminating the need for it altogether

Reply:
Please note that the pointer was present in the old `FlowController.h:343` and the members at old `FlowController.cpp:363-522`.

I wouldn't support introducing a pimpl here, but I support the extraction of this existing pimpl to its own class, because it means one less responsibility for the bloated `FlowController` and because it's a meaningful class on its own.

I also support eliminating it altogether, but it might require more changes and I don't think it's a good idea to further inflate this PR.

Reply:
Okay, I like this approach.

Let's leave this PR as it is and create a follow-up to find a better solution later.